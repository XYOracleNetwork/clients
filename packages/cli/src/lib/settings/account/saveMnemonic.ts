import { saveSettings } from '../saveSettings.js'
import type { Settings } from '../Settings.js'

export const saveMnemonic = async (mnemonic: string): Promise<void> => {
  const data: Settings = { account: { mnemonic } }
  await saveSettings(data)
}
