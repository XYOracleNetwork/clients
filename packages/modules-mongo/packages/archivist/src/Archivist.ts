import { exists } from '@xylabs/exists'
import { Hash } from '@xylabs/hex'
import { AbstractArchivist } from '@xyo-network/archivist-abstract'
import { ArchivistConfigSchema, ArchivistInsertQuerySchema } from '@xyo-network/archivist-model'
import { MongoDBArchivistConfigSchema } from '@xyo-network/archivist-model-mongodb'
import { MongoDBModuleMixin } from '@xyo-network/module-abstract-mongodb'
import { PayloadBuilder } from '@xyo-network/payload-builder'
import { Payload, WithMeta } from '@xyo-network/payload-model'
import { fromDbRepresentation, PayloadWithMongoMeta, toDbRepresentation } from '@xyo-network/payload-mongodb'
import { PayloadWrapper } from '@xyo-network/payload-wrapper'

import { validByType } from './lib'

const MongoDBArchivistBase = MongoDBModuleMixin(AbstractArchivist)

export class MongoDBArchivist extends MongoDBArchivistBase {
  static override configSchemas = [MongoDBArchivistConfigSchema, ArchivistConfigSchema]

  override readonly queries: string[] = [ArchivistInsertQuerySchema, ...super.queries]

  override async head(): Promise<Payload | undefined> {
    const head = await (await this.payloads.find({})).sort({ _timestamp: -1 }).limit(1).toArray()
    return head[0] ? PayloadWrapper.wrap(head[0]).payload : undefined
  }

  protected override async getHandler(hashes: string[]): Promise<WithMeta<Payload>[]> {
    let remainingHashes = [...hashes]

    const dataPayloads = (await Promise.all(remainingHashes.map((_$hash) => this.payloads.findOne({ _$hash })))).filter(exists)
    const dataPayloadsHashes = new Set(dataPayloads.map((payload) => payload._$hash))
    remainingHashes = remainingHashes.filter((hash) => !dataPayloadsHashes.has(hash))

    const dataBws = (await Promise.all(remainingHashes.map((_$hash) => this.boundWitnesses.findOne({ _$hash })))).filter(exists)
    const dataBwsHashes = new Set(dataBws.map((payload) => payload._$hash))
    remainingHashes = remainingHashes.filter((hash) => !dataBwsHashes.has(hash))

    const payloads = (await Promise.all(remainingHashes.map((_hash) => this.payloads.findOne({ _hash })))).filter(exists)
    const payloadsHashes = new Set(payloads.map((payload) => payload._hash))
    remainingHashes = remainingHashes.filter((hash) => !payloadsHashes.has(hash))

    const bws = (await Promise.all(remainingHashes.map((_hash) => this.boundWitnesses.findOne({ _hash })))).filter(exists)
    const bwsHashes = new Set(bws.map((payload) => payload._hash))
    remainingHashes = remainingHashes.filter((hash) => !bwsHashes.has(hash))

    const foundPayloads = [...dataPayloads, ...dataBws, ...payloads, ...bws] as PayloadWithMongoMeta<Payload & { _$hash: Hash; _$meta?: unknown }>[]
    return await PayloadBuilder.build(foundPayloads.map(fromDbRepresentation))
  }

  protected override async insertHandler(payloads: Payload[]): Promise<WithMeta<Payload>[]> {
    const [bw, p] = await validByType(payloads)
    const payloadsWithExternalMeta = await Promise.all(p.map(toDbRepresentation))
    if (payloadsWithExternalMeta.length > 0) {
      const payloadsResult = await this.payloads.insertMany(payloadsWithExternalMeta)
      if (!payloadsResult.acknowledged || payloadsResult.insertedCount !== payloadsWithExternalMeta.length)
        throw new Error('MongoDBDeterministicArchivist: Error inserting Payloads')
    }
    const boundWitnessesWithExternalMeta = await Promise.all(bw.map(toDbRepresentation))
    if (boundWitnessesWithExternalMeta.length > 0) {
      const boundWitnessesResult = await this.boundWitnesses.insertMany(boundWitnessesWithExternalMeta)
      if (!boundWitnessesResult.acknowledged || boundWitnessesResult.insertedCount !== boundWitnessesWithExternalMeta.length)
        throw new Error('MongoDBDeterministicArchivist: Error inserting BoundWitnesses')
    }

    return await PayloadBuilder.build([...boundWitnessesWithExternalMeta, ...payloadsWithExternalMeta])
  }

  protected override async startHandler() {
    await super.startHandler()
    await this.ensureIndexes()
    return true
  }
}
