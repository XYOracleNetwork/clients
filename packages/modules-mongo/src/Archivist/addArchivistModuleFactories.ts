import { MongoDBArchivist } from '@xyo-network/archivist-mongodb'
import type { ModuleFactoryLocator } from '@xyo-network/module-factory-locator'
import { TYPES } from '@xyo-network/node-core-types'
import type { Container } from 'inversify'

export const addArchivistModuleFactories = (container: Container) => {
  const locator = container.get<ModuleFactoryLocator>(TYPES.ModuleFactoryLocator)
  locator.register(MongoDBArchivist.factory())
}
