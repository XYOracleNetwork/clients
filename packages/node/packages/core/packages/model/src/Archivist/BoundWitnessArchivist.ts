import type { Archivist } from '@xyo-network/archivist-model'
import type { BoundWitness } from '@xyo-network/boundwitness-model'
import type { AbstractModule } from '@xyo-network/module-abstract'

export type BoundWitnessArchivist = Archivist<BoundWitness, BoundWitness, BoundWitness, string> & AbstractModule
