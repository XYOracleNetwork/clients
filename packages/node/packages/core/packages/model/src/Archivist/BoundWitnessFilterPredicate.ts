import type { BoundWitnessWithMongoMeta, PayloadWithMongoMeta } from '@xyo-network/payload-mongodb'

import type { PayloadFilterPredicate } from './PayloadFilterPredicate.js'

export type WithoutSchemas<T> = Omit<Omit<T, 'schema'>, 'schemas'>

// TODO: Should we just accept "schema"/"schemas" here and infer that they mean "payload_schemas"?
export type BoundWitnessFilterPredicate = WithoutSchemas<PayloadFilterPredicate> & Partial<BoundWitnessWithMongoMeta & PayloadWithMongoMeta>
