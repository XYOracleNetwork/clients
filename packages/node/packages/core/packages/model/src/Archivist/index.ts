export * from './ArchiveModuleConfig.js'
export * from './BoundWitnessArchivist.js'
export * from './BoundWitnessesArchivist.js'
export * from './BoundWitnessFilterPredicate.js'
export * from './PayloadFilterPredicate.js'
