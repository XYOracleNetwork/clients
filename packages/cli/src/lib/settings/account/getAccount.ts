import { generateMnemonic } from '@scure/bip39'

import { wordlist } from '@scure/bip39/wordlists/english'
import { HDWallet } from '@xyo-network/account'
import { AccountInstance } from '@xyo-network/account-model'

import { loadMnemonic } from './loadMnemonic'
import { saveMnemonic } from './saveMnemonic'

export const getAccount = async (): Promise<AccountInstance> => {
  let mnemonic = await loadMnemonic()
  if (!mnemonic) {
    mnemonic = generateMnemonic(wordlist, 256)
    await saveMnemonic(mnemonic)
  }
  const account = HDWallet.fromPhrase(mnemonic)
  return account
}
