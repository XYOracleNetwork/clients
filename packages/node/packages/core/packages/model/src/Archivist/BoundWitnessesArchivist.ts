import { BoundWitnessArchivist } from './BoundWitnessArchivist.js'

export type BoundWitnessesArchivist = BoundWitnessArchivist
