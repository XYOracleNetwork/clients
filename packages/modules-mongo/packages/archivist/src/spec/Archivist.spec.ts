import { describeIf } from '@xylabs/jest-helpers'
import { Account } from '@xyo-network/account'
import { ArchivistWrapper } from '@xyo-network/archivist-wrapper'
import { BoundWitnessBuilder } from '@xyo-network/boundwitness-builder'
import { BoundWitnessWrapper } from '@xyo-network/boundwitness-wrapper'
import { COLLECTIONS, hasMongoDBConfig } from '@xyo-network/module-abstract-mongodb'
import { PayloadBuilder } from '@xyo-network/payload-builder'
import type { PayloadWrapperBase } from '@xyo-network/payload-wrapper'
import { PayloadWrapper } from '@xyo-network/payload-wrapper'
import type { BaseMongoSdkConfig } from '@xyo-network/sdk-xyo-mongo-js'

import { MongoDBArchivist } from '../Archivist.js'

describeIf(hasMongoDBConfig())('DeterministicArchivist', () => {
  const boundWitnessesConfig: BaseMongoSdkConfig = { collection: COLLECTIONS.BoundWitnesses }
  const payloadsConfig: BaseMongoSdkConfig = { collection: COLLECTIONS.Payloads }

  const payloadWrappers: PayloadWrapper[] = []
  const boundWitnessWrappers: BoundWitnessWrapper[] = []
  let archivist: ArchivistWrapper
  beforeAll(async () => {
    boundWitnessesConfig.dbConnectionString = process.env.MONGO_CONNECTION_STRING
    payloadsConfig.dbConnectionString = process.env.MONGO_CONNECTION_STRING

    const mod = await MongoDBArchivist.create({
      account: 'random',
      boundWitnessSdkConfig: boundWitnessesConfig,
      config: { schema: MongoDBArchivist.defaultConfigSchema },
      payloadSdkConfig: payloadsConfig,
    })
    archivist = new ArchivistWrapper({ mod: mod, account: await Account.random() })
    const payload1 = await PayloadBuilder.build({ nonce: 1, schema: 'network.xyo.debug' })
    const payload2 = await PayloadBuilder.build({ nonce: 2, schema: 'network.xyo.test' })
    const payload3 = await PayloadBuilder.build({ nonce: 3, schema: 'network.xyo.debug' })
    const payload4 = await PayloadBuilder.build({ nonce: 4, schema: 'network.xyo.test' })
    const payloadWrapper1 = PayloadWrapper.wrap(payload1)
    const payloadWrapper2 = PayloadWrapper.wrap(payload2)
    const payloadWrapper3 = PayloadWrapper.wrap(payload3)
    const payloadWrapper4 = PayloadWrapper.wrap(payload4)
    payloadWrappers.push(payloadWrapper1, payloadWrapper2, payloadWrapper3, payloadWrapper4)
    const signer = await Account.random()
    const boundWitness1 = (await new BoundWitnessBuilder().payload(payloadWrapper1.payload).signer(signer).build())[0]
    const boundWitness2 = (await new BoundWitnessBuilder().payload(payloadWrapper2.payload).signer(signer).build())[0]
    const boundWitness3 = (
      await new BoundWitnessBuilder().payloads([payloadWrapper3.payload, payloadWrapper4.payload]).signer(signer).build()
    )[0]
    const boundWitnessWrapper1 = BoundWitnessWrapper.parse(boundWitness1, [payload1])
    const boundWitnessWrapper2 = BoundWitnessWrapper.parse(boundWitness2, [payload2])
    const boundWitnessWrapper3 = BoundWitnessWrapper.parse(boundWitness3, [payload3, payload4])
    boundWitnessWrappers.push(boundWitnessWrapper1, boundWitnessWrapper2, boundWitnessWrapper3)
  })

  describe('discover', () => {
    it('discovers module', async () => {
      const result = await archivist.state()
      expect(result).toBeArray()
      expect(result.length).toBeGreaterThan(0)
    })
  })
  describe('insert', () => {
    beforeEach(() => {})
    type TestDataGetter<T> = () => T
    const cases: [string, TestDataGetter<PayloadWrapperBase[]>][] = [
      ['inserts single payload', () => [payloadWrappers[0]]],
      ['inserts multiple payloads', () => [payloadWrappers[0], payloadWrappers[1], payloadWrappers[2]]],
      ['inserts single boundwitness', () => [boundWitnessWrappers[0]]],
      ['inserts multiple boundwitness', () => [boundWitnessWrappers[0], boundWitnessWrappers[1], boundWitnessWrappers[2]]],
    ]
    it.each(cases)('%s', async (_title, getData) => {
      const payloads = getData().map(w => w.payload)
      const results = await archivist.insert(payloads)
      expect(results).toBeArrayOfSize(payloads.length)
    })
  })
  describe('get', () => {
    type TestDataGetter<T> = () => T
    const cases: [string, TestDataGetter<PayloadWrapperBase[]>][] = [
      ['gets single payload', () => [payloadWrappers[0]]],
      ['gets multiple payloads', () => [payloadWrappers[0], payloadWrappers[1], payloadWrappers[2]]],
      ['gets single boundwitness', () => [boundWitnessWrappers[0]]],
      ['gets multiple boundwitness', () => [boundWitnessWrappers[0], boundWitnessWrappers[1], boundWitnessWrappers[2]]],
    ]
    it.each(cases)('%s', async (_title, getData) => {
      const payloads = getData().map(w => w.payload)
      const dataHashes = await PayloadBuilder.dataHashes(payloads)
      const hashes = await PayloadBuilder.hashes(payloads)
      const dataResults = await archivist.get(dataHashes)
      const results = await archivist.get(hashes)
      expect(results).toBeTruthy()
      expect(results).toBeArrayOfSize(payloads.length)
      const resultHashes = await PayloadBuilder.hashes(results)
      const dataResultHashes = await PayloadBuilder.dataHashes(dataResults)
      await Promise.all(
        payloads.map(async (p) => {
          expect(resultHashes).toInclude(await PayloadBuilder.hash(p))
          expect(dataResultHashes).toInclude(await PayloadBuilder.dataHash(p))
        }),
      )
    })
  })
})
