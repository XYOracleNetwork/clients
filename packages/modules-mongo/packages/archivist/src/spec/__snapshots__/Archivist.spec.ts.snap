// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeterministicArchivist get gets multiple boundwitness 1`] = `
[
  {
    "_signatures": [
      "325501822063bf69e34f6a64f96ef1e55e8eed5060c0dc300271be82b4df024163b2b294018c5d06df160b20394f460b955c77d8a5147ec7f1bafa79b4798662",
    ],
    "addresses": [
      "10ca1959336ea208bcdf00dd6d6637aec91a4c0e",
    ],
    "payload_hashes": [
      "15b8d0e30ca5aa96ca6cc9e1528c075aec88cd3f2c3eb0394fde647eb4bf4547",
    ],
    "payload_schemas": [
      "network.xyo.debug",
    ],
    "previous_hashes": [
      null,
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
  {
    "_signatures": [
      "1810b9f9c6623adf38d0093ea6f3c2125570ed1ee757dbe9e9221b4ffb3920ec0b11175fdffc7e0bfee5d844f9e43c7e261d2bfcd1b273b46caa5cb23a4c798d",
    ],
    "addresses": [
      "10ca1959336ea208bcdf00dd6d6637aec91a4c0e",
    ],
    "payload_hashes": [
      "1c9033e467d6877e2c84a70b29212a585bd6fd448bccfcc12adf0505c18984f6",
    ],
    "payload_schemas": [
      "network.xyo.test",
    ],
    "previous_hashes": [
      "58c138dd2aab7bd78bd2c3d80bebf900d3f99f506e201f1f0ace1d473fc45d76",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
  {
    "_signatures": [
      "29a8dfe2cf7675e8c371763d6d8c7d9b0990145191967c6e399fff7854cd22390d6a8a7f4ae7628ac62deea6c34af33c37623950b1aba6ae4b8c88683d615c20",
    ],
    "addresses": [
      "10ca1959336ea208bcdf00dd6d6637aec91a4c0e",
    ],
    "payload_hashes": [
      "1137543b5805a18e8f39cb7debe49ddf58a1de00c46ed7310baf7ebcfdf435a2",
      "aa36451ef0079af6cc2f86eaa05c7a93fb3c92098e18d0ce857e9e3072b32a85",
    ],
    "payload_schemas": [
      "network.xyo.debug",
      "network.xyo.test",
    ],
    "previous_hashes": [
      "c5e4a2513ea22a7ccfa359207b4d83af68688dca8c9a3f5d39106ba6e599b481",
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
]
`;

exports[`DeterministicArchivist get gets multiple payloads 1`] = `
[
  {
    "nonce": 1,
    "schema": "network.xyo.debug",
  },
  {
    "nonce": 2,
    "schema": "network.xyo.test",
  },
  {
    "nonce": 3,
    "schema": "network.xyo.debug",
  },
]
`;

exports[`DeterministicArchivist get gets single boundwitness 1`] = `
[
  {
    "_signatures": [
      "325501822063bf69e34f6a64f96ef1e55e8eed5060c0dc300271be82b4df024163b2b294018c5d06df160b20394f460b955c77d8a5147ec7f1bafa79b4798662",
    ],
    "addresses": [
      "10ca1959336ea208bcdf00dd6d6637aec91a4c0e",
    ],
    "payload_hashes": [
      "15b8d0e30ca5aa96ca6cc9e1528c075aec88cd3f2c3eb0394fde647eb4bf4547",
    ],
    "payload_schemas": [
      "network.xyo.debug",
    ],
    "previous_hashes": [
      null,
    ],
    "schema": "network.xyo.boundwitness",
    "timestamp": 123456789000,
  },
]
`;

exports[`DeterministicArchivist get gets single payload 1`] = `
[
  {
    "nonce": 1,
    "schema": "network.xyo.debug",
  },
]
`;
